server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java

```diff
@@ -942,6 +947,30 @@ protected void parseCreateField(DocumentParserContext context) throws IOExceptio
     }
 
     private void indexValue(DocumentParserContext context, long timestamp) {
+        // DataStreamTimestampFieldMapper and TsidExtractingFieldMapper need to use timestamp value,
+        // so when this is true we store it in a well-known place
+        // instead of forcing them to iterate over all fields.
+        //
+        // DataStreamTimestampFieldMapper is present and enabled both
+        // in data streams and standalone indices in time_series mode
+        if (isDataStreamTimestampField && context.mappingLookup().isDataStreamTimestampFieldEnabled()) {
+            var existingField = context.doc().getByKey(DataStreamTimestampFieldMapper.TIMESTAMP_VALUE_KEY);
+            if (existingField == null) {
+                context.doc()
+                    .onlyAddKey(
+                        DataStreamTimestampFieldMapper.TIMESTAMP_VALUE_KEY,
+                        new LongField(DataStreamTimestampFieldMapper.TIMESTAMP_VALUE_KEY, timestamp, Field.Store.NO)
+                    );
+            } else if (context.doc().getByKey(DataStreamTimestampFieldMapper.TIMESTAMP_MULTIPLE_VALUES_FLAG) == null) {
```

Comment:
Maybe instead of another flag to deal with multiple timestamps, maybe just fail when there is a value for `DataStreamTimestampFieldMapper.TIMESTAMP_VALUE_KEY` when second timestamp is encountered?

Reply:
My idea behind this is to have all validation logic for `@timestamp` inside `DataStreamTimestampFieldMapper` but i don't have a strong opinion. Let me know what you think.

Reply:
I think we can keep the validation here? For example:

```java
var existingField = context.doc().getByKey(DataStreamTimestampFieldMapper.TIMESTAMP_VALUE_KEY);
if (existingField == null) {
   context.doc().onlyAddKey(DataStreamTimestampFieldMapper.TIMESTAMP_VALUE_KEY, new LongField(DataStreamTimestampFieldMapper.TIMESTAMP_VALUE_KEY, timestamp, Field.Store.NO));
} else {
   throw new IllegalArgumentException("data stream timestamp field [" + DEFAULT_PATH + "] encountered multiple values");
}
```

Reply:
To be clear, my concern is that some validation is done inside `DataStreamTimestampFieldMapper` and some would be done inside `DateFieldMapper`. That being said they are already pretty entangled. I'll rework this a bit.